File: ruby/day12_part1_2015.rb

require 'json'

def sum_numbers(data)
  case data
  when Array
    data.sum { |v| sum_numbers(v) }
  when Hash
    data.values.sum { |v| sum_numbers(v) }
  when Numeric
    data.to_i
  else
    0
  end
end

json_data = JSON.parse(File.read('input.txt'))
puts sum_numbers(json_data)

File: perl/day12_part1_2015.pl

use JSON;

open(my $fh, "<", "input.txt") or die "Error reading input: $!";
my $data;
{
    local $/;
    $data = <$fh>;
}
close($fh);

my $jsonData = decode_json($data);

my $sum = sumNumbers($jsonData);
print $sum . "\n";

sub sumNumbers {
    my $data = shift;
    my $sum = 0;
    if (ref($data) eq 'ARRAY') {
        foreach my $v (@$data) {
            $sum += sumNumbers($v);
        }
    }
    elsif (ref($data) eq 'HASH') {
        foreach my $v (values %$data) {
            $sum += sumNumbers($v);
        }
    }
    elsif (ref($data) eq '') {
        $sum += int($data);
    }
    return $sum;
}
